program-bpf-rust/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+#![no_std]
+
+extern crate serde;
+#[macro_use]
+extern crate serde_derive;
+#[cfg(not(test))]
+extern crate solana_sdk_bpf_no_std;
+extern crate solana_sdk_bpf_utils;
+
+// required to use vec! in the program_command.h tests
+#[cfg(test)]
+#[macro_use]
+extern crate alloc;
+
+mod dashboard;
+mod game;
+mod program_command;
+mod program_state;
+mod result;
+mod simple_serde;
+
+use program_command::Command;
+use program_state::State;
+use result::{ProgramError, Result as ProgramResult};
+use simple_serde::SimpleSerde;
+use solana_sdk_bpf_utils::entrypoint::*;
+use solana_sdk_bpf_utils::{entrypoint, info};
+
+fn get_current_slot(account: &SolKeyedAccount) -> u64 {
+    #[allow(clippy::cast_ptr_alignment)]
+    unsafe {
+        *(&account.data[0] as *const u8 as *const u64)
+    }
+}
+
+fn expect_n_accounts(info: &mut [SolKeyedAccount], n: usize) -> ProgramResult<()> {
+    if info.len() < n {
+        info!("Incorrect number of accounts");
+        Err(ProgramError::InvalidInput)
+    } else {
+        Ok(())
+    }
+}
+
+fn fund_next_move(
+    info: &mut [SolKeyedAccount],
+    dashboard_index: usize,
+    user_or_game_index: usize,
+) -> ProgramResult<()> {
+    if *info[dashboard_index].lamports <= 1 {
+        info!("Dashboard is out of lamports");
+        Err(ProgramError::InvalidInput)
+    } else {
+        if *info[user_or_game_index].lamports != 0 {
+            info!("User or Game still has lamports");
+        } else {
+            // TODO: the fee to charge may be dynamic based on the FeeCalculator and
+            //       instead probably needs to be passed in as an argument somehow
```

Comment:
While you're here, can you please update this comment to reference https://github.com/solana-labs/solana/blob/50c6b5d62ddb79abbb658388c04a5fd3ded01990/sdk/src/syscall/fees.rs#L14 now that the Fees syscall exists (it didn't when I first wrote that comment)

Reply:
https://github.com/solana-labs/solana/issues/4981